### ISSUE TITLE
audio file format

### ISSUE THREAD

**Issue by satyajitghana** [0x006F/react-media-recorder#31]
Hi,

I am trying to record audio in the browser

```JavaScript
    const {
        status,
        startRecording,
        stopRecording,
        mediaBlobUrl,
    } = useReactMediaRecorder({
        video: false,
        audio: true,
        blobPropertyBag: {
            type: "audio/wav"
        }
    });
```

and fetching the file from it

```JavaScript
            const audioBlob = await fetch(mediaBlobUrl).then(r => r.blob());

            console.log(audioBlob);

            const audiofile = new File([audioBlob], `${uuidv4()}.wav`, { type: "audio/wav" })
```

Now i saved this file in disk, and used `file audio.wav`

and it shows up as as WebM file

```
$ file b61508fa-9c4e-47fc-bb54-5def7abf9bfc.wav                                                         ─╯
b61508fa-9c4e-47fc-bb54-5def7abf9bfc.wav: WebM
```

what i need is a wav file ! and it is giving me a WebM file

**Comment by gdrbyKo1** [comment]
Hi. As far as I can tell, the MediaRecorder API currently doesn't support the `audio/wav` MIME type. This is unrelated to `react-media-recorder`, please see https://github.com/w3c/mediacapture-record/issues/198 and the related Firefox and Chromium issues for details.

**Comment by satyajitghana** [comment]
oh okay, thanks, so probably the `mimeType` has to be changed in your library ? it's kind of misleading.

also i was wondering, can the bytes in the recorded file be changed to `audio/wav`, i.e. can we convert `audio/webm` to `audio/wav` on the client side ?

**Comment by 0x006F** [comment]
`blobPropertyBag` is used only for generating the blobs (_after the recording has been completed_) inside the media recorder. If you want to use a specific codec to use, you need to pass through `mediaRecorderOptions`. It'll check for the browser support first and will emit a `console.error` if there's no support. I'd say not to use the `blobPropertyBag` if it's not a big-deal. It's just a placeholder for extending the `Blob` object. Or else we need to keep them in-sync. 

That being said, currently this library defaults to `video/mp4` /  `audio/wav` while generating the blob, which is kinda incorrect. That needs to be fixed.  

For a quick conversion, I'd suggest `ffmepg`, which has webm --> wav support I believe.

**Comment by satyajitghana** [comment]
yes thanks, ffmpeg worked ! added it on the server side though, and just realised I could have done it on client side instead.

anyways I was able to convert webm to wav

**Comment by liadbelad** [comment]
Hey I got the same issue and need to convert to wav file in client side so I can send a wav file to a server.
How can I implement using ffmpeg in the client side please?

**Comment by 0x006F** [comment]
`ffmpeg`'s usage is pretty straight forward. [Download it from here](https://ffmpeg.org/) and use it like this:  

```bash  
ffmpeg -i <input_webm> -c:a pcm_f32le .<output>.wav  
```

**Comment by satyajitghana** [comment]
@liadbelad see https://www.npmjs.com/package/ffmpeg

**Comment by satyajitghana** [comment]
@liadbelad not sure if this would work, https://www.npmjs.com/package/ffmpeg

**Comment by gdrbyKo1** [comment]
@liadbelad for client-side, there's https://github.com/Kagami/ffmpeg.js/ and https://github.com/ffmpegwasm/ffmpeg.wasm
But are you sure you really want to do that? Sending uncompressed audio data over a network will require significantly more bandwidth. It might be a better choice to do the decoding on the receiving side.

**Comment by startupgurukul** [comment]
Ffmpeg is an overkill..try this https://github.com/ai/audio-recorder-polyfill it has wav encoder and mp3 encoder works in iOS world as well

**Comment by guest271314** [comment]
See https://github.com/guest271314/audioInputToWav/blob/master/index.html

**Comment by guest271314** [comment]
> Hi. As far as I can tell, the MediaRecorder API currently doesn't support the `audio/wav`

Note, Chrome does support `x-matroska;codecs=pcm` where the PCM can be extracted, see https://github.com/WebAudio/web-audio-api-v2/issues/63.

Using Web Audio API a `MediaStream` can be set at `MediaStreamAudioSourceNode` or at `HTMLMediaElement` then connected to an `AudioWorklet` where the `Float32Array`s from `inputs` can be set at an `Array`, `ArrayBuffer`, `SharedArrayBuffer`, or other means of storing data, then WAV headers can be set, e.g., https://github.com/ai/audio-recorder-polyfill/issues/7#issuecomment-744066897.

**Comment by vinikatyal** [comment]
Does this still not work?

**Comment by guest271314** [comment]
This is what I used the last time I needed to convert floats to WAV https://github.com/guest271314/webcodecs/blob/main/WavAudioEncoder.js, usage

```
  const wav = await new WavAudioEncoder({
    sampleRate: 48000,
    numberOfChannels: 1,
    buffers: [new Float32Array(48000)],
  }).encode();
```

**Comment by olegpanchenko** [comment]
```
 useReactMediaRecorder({ mediaRecorderOptions: { mimeType: 'audio/wav' } });
```

**Comment by ramonrovirosa** [comment]
In case it helps anyone in the future, I posted my solution using ffmpeg here https://github.com/0x006F/react-media-recorder/issues/101#issuecomment-1499626166

**Comment by renny-ren** [comment]
> `blobPropertyBag` is used only for generating the blobs (_after the recording has been completed_) inside the media recorder. If you want to use a specific codec to use, you need to pass through `mediaRecorderOptions`. It'll check for the browser support first and will emit a `console.error` if there's no support. I'd say not to use the `blobPropertyBag` if it's not a big-deal. It's just a placeholder for extending the `Blob` object. Or else we need to keep them in-sync.
> 
> That being said, currently this library defaults to `video/mp4` / `audio/wav` while generating the blob, which is kinda incorrect. That needs to be fixed.
> 
> For a quick conversion, I'd suggest `ffmepg`, which has webm --> wav support I believe.

This should be written in the document 😀 I was thought `blobPropertyBag` could do the conversion.

**Comment by dusadvarshit** [comment]
> ```
>  useReactMediaRecorder({ mediaRecorderOptions: { mimeType: 'audio/wav' } });
> ```

This worked for me on Google Chrome

**Comment by Brielle28** [comment]
hello please i need help, I am facing an issue with the download functionality in the application. When I try to download a recording, I receive an "internet connection error" in the download history, and the download does not proceed as expected.

### **here is how i am writing the download functionality**

1. **Download Function:**
   Here’s the `downloadRecording` function that I am using to handle the download:

   ```javascript
   const downloadRecording = (recordingUrl, fileName) => {
       const a = document.createElement("a");
       a.href = recordingUrl;
       a.download = `${fileName}.mp3`; // Use the provided file name
       a.click(); // Trigger the download
   };
   ```

2. **Using the Download Function:**
   In the component where the user can download a recording, I’m using the following code to find the selected recording and call the `downloadRecording` function:

   ```javascript
   const { downloadRecording, recordings } = useContext(UserContext);

   const selectedRecording = recordings.find(
       (recording) => recording.id === audioId
   );

   action.icon === "Download" ? (
       <div
           onClick={() => {
               // Only download if a specific recording is found
               if (selectedRecording) {
                   console.log("Downloading:", {
                       url: selectedRecording.url,
                       name: selectedRecording.name,
                   });
                   downloadRecording(
                       selectedRecording.url,
                       selectedRecording.name
                   );
               } else {
                   console.error("No recording found for ID:", audioId);
                   alert("Recording not found. Please try again.");
               }
           }}
           style={{ cursor: "pointer" }}
       >
           {Content}
       </div>
   );
   ```

### **Expected Behavior:**

The recording should download successfully when the "Download" option is clicked, with the file being saved as `fileName.mp3`.

### **Actual Behavior:**

The download fails with an "internet connection error" message in the download history, and the download does not proceed.

### **Troubleshooting Steps Taken:**

1. I’ve logged the `selectedRecording.url` and `selectedRecording.name` to verify that the correct values are being passed to the `downloadRecording` function.
2. The file does not download, and the error message appears, which suggests there may be an issue with how the blob URL is being handled.

![image](https://github.com/user-attachments/assets/916dced8-4778-463d-94c5-18812e725794)
